internal/engine/engine.go

```diff
@@ -262,24 +262,50 @@ func (engine *Engine) doPlanResources(ctx context.Context, input *enginev1.PlanR
 
 		maps.Copy(auditTrail.EffectivePolicies, policy.GetMeta().GetSourceAttributes())
 	}
-
-	// get the resource policy check
-	rpName, rpVersion, rpScope := engine.policyAttr(input.Resource.Kind, input.Resource.PolicyVersion, input.Resource.Scope)
-	policySet, err = engine.getResourcePolicySet(ctx, rpName, rpVersion, rpScope, opts.LenientScopeSearch())
+	skipResourcePolicies := false
+	rpEvaluator, err := engine.getRolePolicyEvaluator(ctx, opts.evalParams, ppScope, input.Principal.Roles)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to get check for [%s.%s]: %w", rpName, rpVersion, err)
+		return nil, nil, fmt.Errorf("failed to get role policy evaluator: %w", err)
 	}
-
-	if policy := policySet.GetResourcePolicy(); policy != nil {
-		policyEvaluator := planner.ResourcePolicyEvaluator{Policy: policy, Globals: opts.Globals(), SchemaMgr: engine.schemaMgr, NowFn: nowFn}
-		plan, err := policyEvaluator.EvaluateResourcesQueryPlan(ctx, input)
+	if rpEvaluator != nil {
+		tctx := tracer.Start(opts.tracerSink)
+		evalResult, err := PlannerEvaluateRolePolicy(ctx, tctx, rpEvaluator, input)
 		if err != nil {
 			return nil, nil, err
 		}
+		effInfo, ok := evalResult.Effects[input.Action]
+		if !ok {
+			return nil, nil, errors.New("role policy evaluator unexpected result")
+		}
+		effect := effInfo.Effect
+		if effect != effectv1.Effect_EFFECT_ALLOW {
+			skipResourcePolicies = true
```

Comment:
The role policy evaluation logic is rather confusing and [continuing to evolve](https://github.com/cerbos/cerbos-private/issues/1380) (so I'm struggling to reason through it myself), but I don't think this is quite right.

In normal `Check` calls, even `Effect_EFFECT_DENY` effects are evaluated in resource policies now, and can be overridden in the case where a resource policy rule issues an allow for an action _if there is another, independent role for that rule_.

[Haines' comment](https://github.com/cerbos/cerbos-private/issues/1334#issuecomment-2406975989) and the linked issues therein do a good job of providing the context as to why.